vm/src/builtins/range.rs

```diff
@@ -514,20 +546,85 @@ impl PyValue for PyRangeIterator {
 }
 
 #[pyimpl(with(PyIter))]
-impl PyRangeIterator {}
+impl PyRangeIterator {
+    #[pymethod(name = "__length_hint__")]
+    fn length_hint(&self) -> usize {
+        let index = self.index.load();
+        if index < self.length {
+            self.length - index
+        } else {
+            0
+        }
+    }
+
+    #[pymethod(name = "__setstate__")]
+    fn setstate(&self, state: PyObjectRef, vm: &VirtualMachine) -> PyResult<()> {
+        self.index
+            .store(get_state(&BigInt::from(self.length), state, vm)?);
+        Ok(())
+    }
+
+    #[pymethod(magic)]
+    fn reduce(&self, vm: &VirtualMachine) -> PyResult {
+        iterrange_reduce(
+            BigInt::from(self.start),
+            BigInt::from(self.length),
+            BigInt::from(self.step),
+            self.index.load(),
+            vm,
+        )
+    }
+}
 
 impl PyIter for PyRangeIterator {
     fn next(zelf: &PyRef<Self>, vm: &VirtualMachine) -> PyResult {
-        if zelf.index.load() < zelf.length {
-            Ok(vm
-                .ctx
-                .new_int(zelf.start + (zelf.index.fetch_add(1) as isize) * zelf.step))
+        // TODO: In pathological case (index == usize::MAX) this can wrap around
+        // (since fetch_add wraps). This would result in the iterator spinning again
+        // from the beginning.
+        let index = zelf.index.fetch_add(1);
+        if index < zelf.length {
+            Ok(vm.ctx.new_int(zelf.start + (index as isize) * zelf.step))
         } else {
             Err(vm.new_stop_iteration())
         }
     }
 }
 
+fn iterrange_reduce(
+    start: BigInt,
+    length: BigInt,
+    step: BigInt,
+    index: usize,
+    vm: &VirtualMachine,
+) -> PyResult {
+    let iter = vm.get_attribute(vm.builtins.clone(), "iter")?;
+    let stop = start.clone() + length * step.clone();
+    let range = PyRange {
+        start: PyInt::from(start).into_ref(vm),
+        stop: PyInt::from(stop).into_ref(vm),
+        step: PyInt::from(step).into_ref(vm),
+    };
+    Ok(vm.ctx.new_tuple(vec![
+        iter,
+        vm.ctx.new_tuple(vec![range.into_object(vm)]),
+        vm.ctx.new_int(index),
+    ]))
+}
+
+// Silently clips state (i.e index) in range [0, usize::MAX].
+fn get_state(length: &BigInt, state: PyObjectRef, vm: &VirtualMachine) -> PyResult<usize> {
```

Comment:
In rust, I think `get` usually means getter with small cost.

how about `make_state` or `range_state`?